### ISSUE TITLE
[MNG-6117] ${session.parallel} not correctly set

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7329]
**[Richard Raumberger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rraumberger)** opened **[MNG-6117](https://issues.apache.org/jira/browse/MNG-6117?redirect=false)** and commented

Running `mvn help:evaluate -T 3` and resolving the property  `session.parallel` returns `true` when run with Maven 3.0.5.

When running it with 3.3.9, this property resolves to `false`.

```
mvn help:evaluate -T 3
[INFO] Scanning for projects...
[INFO]
[INFO] Using the MultiThreadedBuilder implementation with a thread count of 3
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-help-plugin:2.2:evaluate (default-cli) @ standalone-pom ---
[INFO] No artifact parameter specified, using 'org.apache.maven:standalone-pom:pom:1' as project.
[INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
${session.parallel}
[INFO]
false
```

E.g. Surefire [relies](https://github.com/apache/maven-surefire/blob/fb9b699e27dae905393a5bc4d5c6ae39606c3172/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java#L658) on this property, to check whether Maven is running in parallel mode, as mentioned in the comments of [SUREFIRE-1301](https://issues.apache.org/jira/browse/SUREFIRE-1301).

---

**Affects:** 3.3.9

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=67636510)

**Backported to:** [3.5.0-alpha-1](https://github.com/apache/maven/milestone/92?closed=1)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

The `parallel` flag is indeed never set to `true` when the build is done in parallel.

However, there is a special consideration: the current [`MultiThreadedBuilder`](http://maven.apache.org/components/ref/3.3.9/maven-core/apidocs/org/apache/maven/lifecycle/internal/builder/multithreaded/MultiThreadedBuilder.html) will only build in parallel when the degree of concurrency (determined with `-T`) is at least 2, and when there are at least 2 projects to build. So running `mvn help:evaluate -T 3` on a single project (like the stub project) will still build the project with just 1 thread. In this case, which is the example in the description, I think `session.parallel` should still return `false` (and not `true` like Maven 3.0.5 does, in the given example): even though `-T 3` was set, it is ignored and only one thread is going to be used, resulting is a sequential, and not parallel, build.

To me, the fix here is to have `session.parallel` set to `true` when more than 1 thread are actually used to build. `MultiThreadedBuilder` does not clearly document the number of threads used, but if we agree on this, I'll also add it to reflect that.

**Comment by jira-importer** [comment]
**[Richard Raumberger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rraumberger)** commented

I'm sorry for the misunderstanding. I wasn't aware of the fact, that this special behaviour exists for the `MultiThreadedBuilder`. The behaviour does make sense, though.

A valid example using the correct amount of threads, which avoids the mentioned special behaviour can be found [here](https://cloud.almost-a-blog.net/s/dUmO2mZha5hzqB3/download). Which results in `session.parallel` to be `false` for all threads, when run with `mvn clean test -T 3 | grep isParallel`.
If the same command is executed with Maven 3.0.5 the property is correctly set.

Your fix sounds ok for me.

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

Fixed in [dfb4a7b7f4225bc32385da139679e189a2aeb79a](https://git1-us-west.apache.org/repos/asf?p=maven.git;a=commit;h=dfb4a7b7f4225bc32385da139679e189a2aeb79a).

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Maven 3.4.0 has been dropped. See [this thread](https://mail-archives.apache.org/mod_mbox/maven-dev/201701.mbox/%3CCA%2BnPnMx-e7kGYy3Hp87v8hLGdhp1q%3DtKLx_6QuZ4kGUqHEBGcw%40mail.gmail.com%3E) for more details.

This issue will need to be re-scheduled for a Maven release in the (hopefully near) future.
